File: src/main/kotlin/com/solanteq/solar/edu/pga/queue/MSQueue.java
package com.solanteq.solar.edu.pga.queue;

import java.util.concurrent.atomic.AtomicReference;

public class MSQueue<V> implements Queue<V> {
    private final AtomicReference<Node> Head;
    private final AtomicReference<Node> Tail;

    public MSQueue() {
        final Node dummy = new Node(null);
        this.Head = new AtomicReference<>(dummy);
        this.Tail = new AtomicReference<>(dummy);
    }

    @Override
    public boolean isEmpty() {
        return Head == Tail;
    }

    @Override
    public void enqueue(final V x) {
        final Node node = new Node(x);
        Node tail;
        while (true) {
            tail = this.Tail.get();
            if (tail.next.get() == null) {
                if (tail.next.compareAndSet(null, node)) {
                    break;
                }
            } else {
                this.Tail.compareAndSet(tail, tail.next.get());
            }
        }
        this.Tail.compareAndSet(tail, node);
    }

    @Override
    public V dequeue() {
        while (true) {
            final Node head = this.Head.get();
            final Node tail = this.Tail.get();
            final Node next = head.next.get();
            if (head == this.Head.get()) {
                if (head == tail) {
                    if (next == null) {
                        return null;
                    }
                    this.Tail.compareAndSet(tail, next);
                } else {
                    if (this.Head.compareAndSet(head, next)) {
                        return next.x;
                    }
                }
            }
        }
    }

    @Override
    public V peek() {
        final Node next = this.Head.get().next.get();
        if (next == null) {
            return null;
        }
        return next.x;
    }

    private class Node {
        final V x;
        final AtomicReference<Node> next;

        Node(final V x) {
            this.x = x;
            this.next = new AtomicReference<>(null);
        }
    }
}

File: src/main/kotlin/com/solanteq/solar/edu/pga/queue/MSQueue.kt
package com.solanteq.solar.edu.pga.queue

import java.util.concurrent.atomic.AtomicReference

/**
 * @author gpushkarev
 * @since 4.0.0
 */
class MSQueue<V> : Queue<V> {
    private val head: AtomicReference<Node>
    private val tail: AtomicReference<Node>

    init {
        val dummy = Node(null)
        head = AtomicReference<Node>(dummy)
        tail = AtomicReference<Node>(dummy)
    }

    override fun isEmpty(): Boolean {
        return head === tail
    }

    override fun add(x: V) {
        val node = Node(x)
        var tail: Node
        while (true) {
            tail = this.tail.get()
            if (tail.next.get() == null) {
                if (tail.next.compareAndSet(null, node)) {
                    break
                }
            } else {
                this.tail.compareAndSet(tail, tail.next.get())
            }
        }
        this.tail.compareAndSet(tail, node)
    }

    override fun poll(): V? {
        while (true) {
            val head: Node = head.get()
            val tail: Node = tail.get()
            val next: Node? = head.next.get()
            if (head === tail) {
                if (next == null) {
                    return null
                }
                this.tail.compareAndSet(tail, next)
            } else {
                if (this.head.compareAndSet(head, next)) {
                    return next?.x
                }
            }
        }
    }

    override fun peek(): V? {
        return head.get().next.get()?.x
    }

    private inner class Node(val x: V?) {
        val next: AtomicReference<Node?> = AtomicReference(null)
    }
}
